src/Lib.P2P/PubSub/PubSubService.cs

```diff
@@ -0,0 +1,278 @@
+#region LICENSE
+
+/**
+* Copyright (c) 2019 Catalyst Network
+*
+* This file is part of Catalyst.Node <https://github.com/catalyst-network/Catalyst.Node>
+*
+* Catalyst.Node is free software: you can redistribute it and/or modify
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation, either version 2 of the License, or
+* (at your option) any later version.
+*
+* Catalyst.Node is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Catalyst.Node. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#endregion
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Common.Logging;
+
+namespace Lib.P2P.PubSub
+{
+    /// <summary>
+    ///   A simple pub/sub messaging service that supports
+    ///   multiple message routers.
+    /// </summary>
+    /// <remarks>
+    ///   Relies upon the router(s) to deliver and receive messages from other peers.
+    /// </remarks>
+    public sealed class PubSubService : IPubSubService
+    {
+        private static ILog _log = LogManager.GetLogger(typeof(PubSubService));
+
+        private sealed class TopicHandler
+        {
+            public string Topic;
+            public Action<IPublishedMessage> Handler;
+        }
+
+        private long _nextSequenceNumber;
+        private ConcurrentDictionary<TopicHandler, TopicHandler> _topicHandlers;
+        private readonly MessageTracker _tracker = new MessageTracker();
+
+        // TODO: A general purpose CancellationTokenSource that stops publishing of
+        // messages when this service is stopped.
+
+        /// <summary>
+        ///   The local peer.
+        /// </summary>
+        public Peer LocalPeer { get; set; }
+
+        /// <summary>
+        ///   Sends and receives messages to other peers.
+        /// </summary>
+        public List<IMessageRouter> Routers { get; set; } = new List<IMessageRouter>
+        {
+            new LoopbackRouter()
+        };
+
+        /// <summary>
+        ///   The number of messages that have published.
+        /// </summary>
+        public ulong MesssagesPublished;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Make this field 'private' and encapsulate it in a 'public' property.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)